ciri/api/api.c

```diff
@@ -527,11 +529,113 @@ retcode_t iota_api_init(iota_api_t *const api, node_t *const node,
   return RC_OK;
 }
 
+typedef struct iota_api_http_sess_s {
+  bool valid_api_version;
+  bool valid_content_type;
+  char_buffer_t *request;
+} iota_api_http_sess_t;
+
+static int iota_api_http_header_iter(void *cls, enum MHD_ValueKind kind,
+                                     const char *key, const char *value) {
+  iota_api_http_sess_t *sess = cls;
+
+  if (0 == strcmp(MHD_HTTP_HEADER_CONTENT_TYPE, key)) {
+    sess->valid_content_type = !strcmp("application/json", value);
+  } else if (0 == strcmp("X-IOTA-API-Version", key)) {
+    sess->valid_api_version = !strcmp("1", value);
+  }
+
+  return MHD_YES;
+}
+
+static int iota_api_http_handler(void *cls, struct MHD_Connection *connection,
+                                 const char *url, const char *method,
+                                 const char *version, const char *upload_data,
+                                 size_t *upload_data_size, void **ptr) {
+  int ret;
+  iota_api_t *api = (iota_api_t *)cls;
+  iota_api_http_sess_t *sess = *ptr;
+  struct MHD_Response *response;
+
+  if (strncmp(method, MHD_HTTP_METHOD_POST, 4) != 0) {
+    return MHD_NO;
+  }
+
+  if (sess == NULL) {
+    sess = calloc(1, sizeof(iota_api_http_sess_t));
+    *ptr = sess;
+
+    MHD_get_connection_values(connection, MHD_HEADER_KIND,
+                              iota_api_http_header_iter, sess);
+    return MHD_YES;
+  }
+
+  if (!sess->valid_api_version || !sess->valid_content_type) {
+    ret = MHD_NO;
+    goto cleanup;
+  }
+
+  // While upload_data_size > 0 process upload_data
+  if (*upload_data_size > 0) {
+    if (sess->request == NULL) {
+      sess->request = char_buffer_new();
+      char_buffer_allocate(sess->request, *upload_data_size);
+      strncpy(sess->request->data, upload_data, *upload_data_size);
+    } else {
+      log_error(logger_id,
+                "cIRI HTTP API currently does not support parsing such large "
+                "requests.\n");
+      ret = MHD_NO;
+      goto cleanup;
+    }
+
+    *upload_data_size = 0;
+    return MHD_YES;
+  }
+
+  if (sess->request == NULL) {
+    // POST but no body, so we skip this request
+    ret = MHD_NO;
+    goto cleanup;
+  }
+
+  cJSON *json_obj = cJSON_Parse(sess->request->data);
+  cJSON *json_item = cJSON_GetObjectItemCaseSensitive(json_obj, "command");
+
+  if (!cJSON_IsString(json_item) || (json_item->valuestring == NULL)) {
+    cJSON_Delete(json_obj);
+    ret = MHD_NO;
+    goto cleanup;
+  }
+
+  response = MHD_create_response_from_buffer(strlen(json_item->valuestring),
+                                             json_item->valuestring,
+                                             MHD_RESPMEM_MUST_COPY);
+  MHD_add_response_header(response, MHD_HTTP_HEADER_CONTENT_TYPE,
+                          "application/json");
```

Comment:
I'm not sure i'm understanding the flow of this callback function,
Maybe what's missing here is the routing of the command to the actual command impl?
where is the response being queued to in `MHD_queue_response`? i.e, who's the client?

Reply:
this method is called at least 3 times in total (connection establishing with header parsing, payload parsing (repeatedly), final processing)
it's a weird impl quirk of libmicrohttpd